### PR TITLE
[develop < T0086-GA] Change arguments in RETURN method

### PR DESCRIPTION
**Pull request by katarinasupe**
### Description

I've changed arguments in `RETURN`, `YIELD` and `WITH`methods in the query builder according to the discussion in one of the issues: https://github.com/memgraph/gqlalchemy/issues/140. I created a new class called `ResultPartialQuery` from which `ReturnPartialQuery`, `YieldPartialQuery` and `WithPartialQuery` inherit.

### Pull request type

- [x] Code style update (formatting, renaming)
- [x] Refactoring with functional or API changes

### Related issues

Closes https://github.com/memgraph/gqlalchemy/issues/140 and https://github.com/memgraph/gqlalchemy/issues/125

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

######################################

### Reviewer checklist (the reviewer checks this part)
- [ ] Core feature implementation
- [ ] Tests
- [ ] Code documentation
- [ ] Documentation on [memgraph/docs](https://github.com/memgraph/docs)

######################################

### CODE REVIEW

**Review by Josipmrden — APPROVED** [review]

**Comment by katarinasupe** [comment]
As I mentioned in the PR description, I think that `UNWIND`, `YIELD` and `WITH` should be changed in the same way. If you agree, should I do that in this PR or in separate PRs? @g-despot @brunos252 @BorisTasevski @Josipmrden

**Review by g-despot — CHANGES_REQUESTED** [review]

File: gqlalchemy/exceptions.py [hunk 1]
```diff
@@ -119,3 +127,15 @@ class GQLAlchemyExtraKeywordArgumentsInWhere(GQLAlchemyError):
     def __init__(self):
         super().__init__()
         self.message = EXTRA_KEYWORD_ARGUMENTS_IN_WHERE
+
+
+class GQLAlchemyMissingAliasInReturn(GQLAlchemyError):
+    def __init__(self):
+        super().__init__()
```
Comment (BorisTasevski) [inline gqlalchemy/exceptions.py:134]: remove parameter-less base constructor invocation

File: gqlalchemy/exceptions.py [hunk 2]
```diff
@@ -119,3 +127,15 @@ class GQLAlchemyExtraKeywordArgumentsInWhere(GQLAlchemyError):
...
+        self.message = MISSING_ALIAS_IN_RETURN
+
+
+class GQLAlchemyReturnTypeError(TypeError):
+    def __init__(self):
+        super().__init__()
```
Comment (BorisTasevski) [inline gqlalchemy/exceptions.py:140]: remove parameter-less base constructor invocation

File: gqlalchemy/query_builder.py [hunk 3]
```diff
@@ -373,20 +376,55 @@ def construct_query(self) -> str:
...
+
+    def _return_read_item(self, item: Union[str, Tuple[str, str]]) -> str:
+        if isinstance(item, str):
+            return f"{self._return_read_str(item)}"
+        elif isinstance(item, tuple):
+            return f"{self._return_read_tuple(item)}"
+        else:
+            raise GQLAlchemyReturnTypeError
+
+    def _return_read_list(self, results: List[Union[str, Tuple[str, str]]]):
```
Comment (BorisTasevski) [inline gqlalchemy/query_builder.py:408]: List -> Iterable

File: gqlalchemy/query_builder.py [hunk 4]
```diff
@@ -373,20 +376,55 @@ def construct_query(self) -> str:
...
+        return ", ".join(self._return_read_item(item=item) for item in results)
+
+    def _return_read_str(self, item: str) -> str:
```
Comment (BorisTasevski) [inline gqlalchemy/query_builder.py:411]: Why... This... ?

File: gqlalchemy/query_builder.py [hunk 5]
```diff
@@ -373,20 +376,55 @@ def construct_query(self) -> str:
...
+        return f"{item}"
+
+    def _return_read_tuple(self, tuple: Tuple[str, str]) -> str:
+        if not isinstance(tuple[1], str):
+            raise GQLAlchemyMissingAliasInReturn
+
+        if tuple[0] == tuple[1]:
```
Comment (BorisTasevski) [inline gqlalchemy/query_builder.py:418]: if tuple[0] == tuple[1] or tuple[1] == "":
            return f"{tuple[0]}"

File: gqlalchemy/query_builder.py [hunk 6]
```diff
@@ -918,17 +956,18 @@ def yield_(self, results: Optional[Dict[str, str]] = {}) -> "DeclarativeBase":
 
         return self
 
-    def return_(self, results: Optional[Dict[str, str]] = {}) -> "DeclarativeBase":
+    def return_(
+        self, results: Optional[Union[str, Tuple[str, str], List[Union[str, Tuple[str, str]]]]] = None
```
Comment (BorisTasevski) [inline gqlalchemy/query_builder.py:960]: List -> Iterable

**Review by BorisTasevski — CHANGES_REQUESTED** [review]

File: gqlalchemy/query_builder.py [hunk 7]
```diff
@@ -373,20 +375,55 @@ def construct_query(self) -> str:
 
 
 class ReturnPartialQuery(PartialQuery):
-    def __init__(self, results: Dict[str, str]):
+    def __init__(
+        self, results: Optional[Union[str, Tuple[str, str], List[Union[str, Tuple[str, str]]], Dict[str, str]]] = None
```
[the remark this answers is not in the document]
Comment (katarinasupe) [reply gqlalchemy/query_builder.py:380]: Yep, totally forgot about that in this PR. Thank you for mentioning it, I will address this in my next commit.

File: gqlalchemy/exceptions.py (hunk 1 above)
Comment (katarinasupe) [reply gqlalchemy/exceptions.py:134]: Did you mean that I should delete line 134 - `super().__init__()`?
Should I then delete it on other constructors as well?

File: gqlalchemy/query_builder.py (hunk 4 above)
Comment (katarinasupe) [reply gqlalchemy/query_builder.py:411]: Expand your comment please

Comment (katarinasupe) [reply gqlalchemy/query_builder.py]: We discussed this, this method is not necessary and it was used in the `OrderByPartialQuery`. Now it has been removed from there and `ReturnPartialQuery`.

**Comment by katarinasupe** [comment]
Still missing a fix for `Iterable` to work. I will update it tomorrow. Everything else is implemented.

**Review by Josipmrden — APPROVED** [review]

**Review by g-despot — APPROVED** [review]

**Pull request merged**